src/components/utils/ZoomImage.tsx

```diff
@@ -0,0 +1,259 @@
+
+        return () => {
+            resizeObserver?.disconnect();
+            mediaQueryList.removeEventListener('change', onChange);
+        };
+    }, [imgRef, width, setZoomable]);
+
+    // Preload the image that will be displayed in the modal
+    if (zoomable) {
+        ReactDOM.preload(src, {
+            as: 'image',
+        });
+    }
+    const preloadImage = React.useCallback(
+        (onLoad?: () => void) => {
+            const image = new Image();
+            image.src = src;
+
+            image.onload = () => {
+                onLoad?.();
+            };
+        },
+        [src],
+    );
+
+    // When closing the modal, animate the transition back to the original image
+    const onClose = React.useCallback(() => {
+        startViewTransition(
+            () => {
+                setOpened(false);
+            },
+            () => {
+                setActive(false);
+            },
+        );
+    }, [setOpened]);
+
+    return (
+        <>
+            {opened ? (
+                ReactDOM.createPortal(
+                    <ZoomImageModal src={src} alt={alt ?? ''} onClose={onClose} />,
+                    document.body,
+                )
+            ) : (
+                <img
+                    ref={imgRef}
+                    {...props}
+                    alt={alt ?? ''}
+                    onMouseEnter={() => {
+                        if (zoomable) {
+                            preloadImage();
+                        }
+                    }}
+                    onClick={() => {
+                        if (!zoomable) {
+                            return;
+                        }
+
+                        // Preload the image before opening the modal to ensure the animation is smooth
+                        preloadImage(() => {
+                            const change = () => {
+                                setOpened(true);
+                            };
+
+                            ReactDOM.flushSync(() => setActive(true));
+                            startViewTransition(change);
+                        });
+                    }}
+                    className={classNames(
+                        props.className,
+                        zoomable ? styles.zoomImg : null,
+                        active ? styles.zoomImageActive : null,
+                    )}
+                />
+            )}
+        </>
+    );
+}
+
+function ZoomImageModal(props: { src: string; alt: string; onClose: () => void }) {
+    const { src, alt, onClose } = props;
+
+    const buttonRef = React.useRef<HTMLButtonElement>(null);
+
+    React.useEffect(() => {
+        const handleKeyDown = (event: KeyboardEvent) => {
+            if (event.key === 'Escape') {
+                onClose();
+            }
+        };
+
+        document.addEventListener('keydown', handleKeyDown);
+
+        return () => {
+            document.removeEventListener('keydown', handleKeyDown);
+        };
+    }, [onClose]);
```

Comment:
It's stable because contained in the same component, yep but still it's hard to know. You should probably wrap `ZoomImageModal` in a `memo` then we can guess that it awaits stable props.